src/v3/src/components/CaptchaContainer/CaptchaContainer.tsx

```diff
@@ -0,0 +1,135 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import HCaptcha from '@hcaptcha/react-hcaptcha';
+import { Box } from '@okta/odyssey-react-mui';
+import { useEffect, useRef } from 'preact/hooks';
+import ReCAPTCHA from 'react-google-recaptcha';
+
+import { useWidgetContext } from '../../contexts';
+import { useOnSubmit } from '../../hooks';
+import {
+  CaptchaContainerElement,
+  UISchemaElementComponent,
+} from '../../types';
+
+const CaptchaContainer: UISchemaElementComponent<{
+  uischema: CaptchaContainerElement
+}> = ({ uischema }) => {
+  const {
+    options: {
+      siteKey,
+      type: captchaType,
+      captchaId,
+    },
+  } = uischema;
+
+  const { dataSchemaRef } = useWidgetContext();
+  const onSubmitHandler = useOnSubmit();
+  const dataSchema = dataSchemaRef.current!;
+  const captchaRef = useRef<ReCAPTCHA | HCaptcha>(null);
+
+  useEffect(() => {
+    // set the reference in dataSchema context to this captcha instance
+    dataSchema.captchaRef = captchaRef;
+    //  For some reason Hcaptcha does not initially auto render only when in development mode
+    //  This delayed manual render is only used in development mode for hcaptcha
+    if (process.env.NODE_ENV === 'development' && captchaType === 'HCAPTCHA') {
+      setTimeout(() => {
+        (captchaRef.current as HCaptcha)?.renderCaptcha();
+      }, 500);
+    }
+    return () => {
+      dataSchema.captchaRef = undefined;
+    };
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [dataSchema]);
+
+  const onErrorCaptcha = (e: Event | string) => {
+    console.error(e);
+  };
+
+  const resetCaptchaContainer = () => {
+    if (captchaType === 'RECAPTCHA_V2') {
+      (captchaRef.current as ReCAPTCHA)?.reset();
+    }
+    (captchaRef.current as HCaptcha)?.resetCaptcha();
+  };
+
+  const onVerifyCaptcha = (token: string | null) => {
+    if (!token) {
+      console.error('captcha token expired');
+      return;
+    }
+
+    const {
+      submit: {
+        actionParams: params,
+        step,
+        includeImmutableData,
+      },
+    } = dataSchema;
+
+    // insert received captcha success token into the captchaVerify submit param
+    const captchaSubmitParams = {
+      captchaVerify: {
+        captchaToken: token,
+        captchaId,
+      },
+      ...params,
+    };
+
+    // Captcha challenge solved successfully, submit form
+    onSubmitHandler({
+      includeData: true,
+      includeImmutableData,
+      params: captchaSubmitParams,
+      step,
+    });
+
+    // tokens are one-time use, so if we submit the same token twice then it will be rejected
+    // by the server the second time, so we must reset the captcha state
+    resetCaptchaContainer();
```

Comment:
Does this happen while we wait for the form submission? If so, could the captcha get reset and be visible to the user before the next form view loads?

Reply:
this does happen while the form submission is happening, but resetting captcha does not have any visual effect.  Its only behind the scenes, for when the form submit fails due to an api error (eg. wrong password), so that captcha is able to get a fresh token for the next attempt.  Once the next form loads this captcha component is removed from the form.